[clang-tidy] Improve documentation of bugprone-pointer-arithmetic-on-polymorphic-object

MainakSil opened the pull request:
This pull request addresses issue #107831 by improving the documentation for the `bugprone-pointer-arithmetic-on-polymorphic-object` Clang-Tidy check.

### Changes Made:
1. **Fixed Syntax Error**: Corrected the syntax for the virtual destructor in the `Base` class (`virtual ~Base();` instead of `virtual void ~Base();`).
2. **Improved Example**: Replaced the current example with a more accurate one, showing how pointer arithmetic on an array of derived objects can trigger a warning due to potential undefined behavior.
3. **Added Final Keyword Example**: Introduced an additional example where marking the derived class as `final` suppresses the warning, as further inheritance is disallowed.
4. **Clarified Documentation**: Added comments explaining how making a class `final` eliminates the risk of polymorphic pointer arithmetic and suppresses the warning.

This PR improves the clarity of the documentation and provides more relevant examples to help users better understand the behavior of this check.

MainakSil committed:
Improve documentation for bugprone-pointer-arithmetic-on-polymorphic-object

chrchr-github commented:
Thank for addressing this.
I still wonder why using `new`/`delete` with additional UB (deleting pointer with offset) is necessary in the examples.
Why not simply define a function taking a pointer argument and do the arithmetic on that?

MainakSil committed:
Merge branch 'llvm:main' into ud

MainakSil committed:
Merge branch 'llvm:main' into ud

MainakSil committed:
Merge branch 'main' into ud

MainakSil committed:
Update example to remove new/delete and focus on pointer arithmetic

MainakSil committed:
Merge branch 'main' into ud

MainakSil commented:
I have updated the example based on your feedback. The `new`/`delete` usage has been removed to avoid additional undefined behavior. The example now directly passes a pointer to a function, which performs the pointer arithmetic, focusing solely on the issue of undefined behavior in pointer arithmetic on polymorphic objects.

nicovank commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst:
```diff
@@ -50,14 +65,19 @@ Options
     void bar() {
       Base *b = new Base[10];
```
Also pass `b` and `d` as parameters instead of `new`/`delete`.

nicovank commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst:
```diff
@@ -50,14 +65,19 @@ Options
-  
+
       delete[] d;
+
+      FinalDerived *f = new FinalDerived[10];
+      f += 1; // no warning, FinalDerived is final and cannot be further derived
+
+      delete[] f;
```
```suggestion
```

Not relevant to the option.

nicovank reviewed (commented):
Thanks! Some whitespace and comment nits.

MainakSil committed:
Merge branch 'llvm:main' into ud

MainakSil committed:
Address maintainer feedback: remove redundant comments, new/delete, and final example

MainakSil commented:
I have addressed the feedback:
1. Removed unnecessary whitespace.
2. Removed redundant comments and the `FinalDerived` example.
3. Replaced `new/delete` with direct array passing.

Please let me know if further changes are needed.

nicovank commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/pointer-arithmetic-on-polymorphic-object.rst:
```diff
@@ -19,20 +19,23 @@ Example:
+    virtual ~Base();
+    int i;
   };
-
```
Sorry I wasn't clear, I meant to keep the spacing here between structs/functions, but when there is a blank line to not have any whitespace characters on it ('' vs '  '). A couple blanks lines in this first block between structs/functions and LGTM!

MainakSil replied:
Do you mean like this?

![image](https://github.com/user-attachments/assets/9ea57dc0-0820-4147-bf08-ec3bb96f704a)

MainakSil replied:
So should I make any changes?

nicovank replied:
Yes, I like the blank lines between functions/structs like in your screenshot above. Current PR version has structs and functions with no breathing room. Then good to ship.

MainakSil replied:
Check it please.

nicovank replied:
LGTM! Sorry for confusion. Do you need me to click the merge button?

MainakSil committed:
Address maintainer feedback: whitespace

MainakSil commented:
If it looks good to you, please do merge it.

nicovank commented:
![image](https://github.com/user-attachments/assets/a40849e3-435c-4812-8aa0-3fa9ba5b6c5d)

Looks like you have the "private email" setting enabled on GitHub, [I believe LLVM policy is to discourage it](https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it/74223/73). Could you please uncheck the private email box at https://github.com/settings/emails?

Hopefully this is enough to fix it, you might also have to squash and rebase commits, let's see. Thanks!

MainakSil commented:
Unchecked the private email box.

nicovank commented:
Worked. Thanks!
